src/dns.rs

```diff
@@ -0,0 +1,131 @@
+use std::future::Future;
+use std::io;
+use std::net::SocketAddr;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{self, Poll};
+
+use backoff::{backoff::Backoff, exponential::ExponentialBackoff, SystemClock};
+use hyper::client::connect::dns as hyper_dns;
+use hyper::service::Service;
+use once_cell::sync::Lazy;
+use tokio::sync::Mutex;
+use trust_dns_resolver::{
+    config::{ResolverConfig, ResolverOpts},
+    lookup_ip::LookupIpIntoIter,
+    system_conf, AsyncResolver, TokioConnection, TokioConnectionProvider, TokioHandle,
+};
+
+struct ResolverInner {
+    resolver: AsyncResolver<TokioConnection, TokioConnectionProvider>,
+    backoff: ExponentialBackoff<SystemClock>,
+}
+
+type SharedResolver = Arc<Mutex<ResolverInner>>;
+
+static SYSTEM_CONF: Lazy<io::Result<(ResolverConfig, ResolverOpts)>> =
+    Lazy::new(|| system_conf::read_system_conf().map_err(io::Error::from));
+
+#[derive(Clone)]
+pub(crate) struct TrustDnsResolver {
+    state: Arc<Mutex<State>>,
+}
+
+pub(crate) struct SocketAddrs {
+    iter: LookupIpIntoIter,
+}
+
+#[derive(Clone)]
+enum State {
+    Init(Option<ExponentialBackoff<SystemClock>>),
+    Ready(SharedResolver),
+}
+
+impl TrustDnsResolver {
+    pub(crate) fn new() -> io::Result<Self> {
+        SYSTEM_CONF.as_ref().map_err(|e| {
```

Comment:
does this introduce linux distro dependency?

Reply:
interpretation and actual behavior distro system specific. resolv.conf can be dynamically updated etc.